### DISCUSSION TITLE
ORM session using python 'with' context: not getting implicit commit when block is exited

### DISCUSSION THREAD

**Discussion by taigit** [sqlalchemy/sqlalchemy#10436] [Usage Questions] [answered]
This is a follow-on to  #10420 , but I'm ignoring the postgres upgrade in this question.

I want my code to commit when it exits the `with `context block.
I need that in order to make sure my legacy code works w/o modification.

Attached is a test file.
Toggle the if statement on line 35.

if i set the if to 0, then there is no explicit commit and the code reports an error and no row appears in the db.

if i set the if to 1, then a row is inserted.

because of: 
https://docs.sqlalchemy.org/en/20/orm/session_basics.html#using-a-sessionmaker
I expected that setting if to 0 should still work, but it does not.
what else can I check or change to get it to work?

[sandbox_test_code - Copy for bug report.py.txt](https://github.com/sqlalchemy/sqlalchemy/files/12833779/sandbox_test_code.-.Copy.for.bug.report.py.txt)

**Comment by CaselIT** [comment, marked as answer]
Hi,

This seems like the behaviour sqlalchemy has always had if you don't use `begin()`. https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.sessionmaker.begin
Does that example works for you on 1.4?

This would have been better as a message in the other discussion.
Also this is a copy of the file to avoid downloading it
```py
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
##from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import declarative_base
from sqlalchemy import Column, Integer, String
from datetime import datetime

# Create the engine and session
##engine = create_engine('your_database_connection_string')
##engine = create_engine('postgresql+psycopg2://username:password@localhost/mydatabase')
engine = create_engine('')
Session = sessionmaker(bind=engine)
Base = declarative_base()

# Define your table model
class SandboxTestTable(Base):
    __tablename__ = 'sandbox_test_table'
    id = Column(Integer, primary_key=True)
    name = Column(String)
    timestamp = Column(String)

# Create the table if it doesn't exist
Base.metadata.create_all(engine)

# Create a new record with timestamp
timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
new_record = SandboxTestTable(name=f'John Doe - {timestamp}', timestamp=timestamp)

# Print the timestamp
print("Current timestamp:", timestamp)

# Insert the record within a 'with' context
with Session() as session:
    session.add(new_record)
    if 0:
        session.commit()

# Query the inserted row using the timestamp
with Session() as session:
    try:
        inserted_row = session.query(SandboxTestTable).filter_by(timestamp=timestamp).first()
        print("Inserted row:", inserted_row.id, inserted_row.name, inserted_row.timestamp)
    except:
        print(f"Error occurred. The commit did not happen for timestamp={timestamp}")



```

**Comment by taigit** [comment]
yup, the problem was  begin()  missing, but made more complex because it happens during db create code so don't feel / see the error till i use the db which is well after the create code was run.

thanks!
